provisionerd/runner.go

```diff
@@ -0,0 +1,872 @@
+				slog.F("resources", msgType.Complete.Resources),
+				slog.F("state_length", len(msgType.Complete.State)),
+			)
+
+			return msgType.Complete.Resources, nil
+		default:
+			return nil, xerrors.Errorf("invalid message type %q received from provisioner",
+				reflect.TypeOf(msg.Type).String())
+		}
+	}
+}
+
+func (r *runner) runTemplateDryRun() (
+	*proto.CompletedJob, *proto.FailedJob) {
+	// Ensure all metadata fields are set as they are all optional for dry-run.
+	metadata := r.job.GetTemplateDryRun().GetMetadata()
+	metadata.WorkspaceTransition = sdkproto.WorkspaceTransition_START
+	if metadata.CoderUrl == "" {
+		metadata.CoderUrl = "http://localhost:3000"
+	}
+	if metadata.WorkspaceName == "" {
+		metadata.WorkspaceName = "dryrun"
+	}
+	metadata.WorkspaceOwner = r.job.UserName
+	if metadata.WorkspaceOwner == "" {
+		metadata.WorkspaceOwner = "dryrunner"
+	}
+	if metadata.WorkspaceId == "" {
+		id, err := uuid.NewRandom()
+		if err != nil {
+			return nil, r.failedJobf("generate random ID: %s", err)
+		}
+		metadata.WorkspaceId = id.String()
+	}
+	if metadata.WorkspaceOwnerId == "" {
+		id, err := uuid.NewRandom()
+		if err != nil {
+			return nil, r.failedJobf("generate random ID: %s", err)
+		}
+		metadata.WorkspaceOwnerId = id.String()
+	}
+
+	// Run the template import provision task since it's already a dry run.
+	resources, err := r.runTemplateImportProvision(
+		r.job.GetTemplateDryRun().GetParameterValues(),
+		metadata,
+	)
+	if err != nil {
+		return nil, r.failedJobf("run dry-run provision job: %s", err)
+	}
+
+	return &proto.CompletedJob{
+		JobId: r.job.JobId,
+		Type: &proto.CompletedJob_TemplateDryRun_{
+			TemplateDryRun: &proto.CompletedJob_TemplateDryRun{
+				Resources: resources,
+			},
+		},
+	}, nil
+}
+
+func (r *runner) runWorkspaceBuild() (
+	*proto.CompletedJob, *proto.FailedJob) {
+	var stage string
+	switch r.job.GetWorkspaceBuild().Metadata.WorkspaceTransition {
+	case sdkproto.WorkspaceTransition_START:
+		stage = "Starting workspace"
+	case sdkproto.WorkspaceTransition_STOP:
+		stage = "Stopping workspace"
+	case sdkproto.WorkspaceTransition_DESTROY:
+		stage = "Destroying workspace"
+	}
+
+	_, err := r.update(r.forceStopContext, &proto.UpdateJobRequest{
+		JobId: r.job.JobId,
+		Logs: []*proto.Log{{
+			Source:    proto.LogSource_PROVISIONER_DAEMON,
+			Level:     sdkproto.LogLevel_INFO,
+			Stage:     stage,
+			CreatedAt: time.Now().UTC().UnixMilli(),
+		}},
+	})
+	if err != nil {
+		return nil, r.failedJobf("write log: %s", err)
+	}
+
+	// use the forceStopContext so that if we attempt to gracefully cancel, the stream will still be available for us
+	// to send the cancel to the provisioner
+	stream, err := r.provisioner.Provision(r.forceStopContext)
+	if err != nil {
+		return nil, r.failedJobf("provision: %s", err)
+	}
+	defer stream.Close()
+	go func() {
+		select {
+		case <-r.forceStopContext.Done():
+			return
+		case <-r.gracefulContext.Done():
+			_ = stream.Send(&sdkproto.Provision_Request{
```

Comment:
Sorry, I kinda mentally skipped over it because that's existing code, other than the names of the contexts.

But yeah, I think your analysis is right, the stream might be closed at the point this case fires, but sending the cancel is best effort anyway.